File: c/src/classical/substitution/polyalphabetic/reihenschieber.c
/**
 * @file reihenschieber.c
 * @brief Reihenschieber Cipher Implementation
 * 
 * A mechanical polyalphabetic substitution cipher that uses shifting strips.
 * Essentially a mechanical Vigenère cipher with progressive shifting capabilities.
 */

#include "../../../../include/cryptology/classical/substitution/polyalphabetic/reihenschieber.h"
#include <string.h>
#include <stdlib.h>
#include <time.h>
#include <ctype.h>

#define DEFAULT_ALPHABET "abcdefghijklmnopqrstuvwxyz"
#define TURKISH_ALPHABET "ABCÇDEFGĞHIİJKLMNOÖPRSŞTUÜVYZ"
#define MAX_ALPHABET_SIZE 50
#define MAX_TEXT_SIZE 1000

/**
 * @brief Convert string to uppercase
 */
static void to_upper(char* str) {
    for (int i = 0; str[i]; i++) {
        str[i] = tolower(str[i]);
    }
}

/**
 * @brief Find character index in alphabet
 */
static int find_char_index(const char* alphabet, char c) {
    for (int i = 0; alphabet[i]; i++) {
        if (alphabet[i] == c) {
            return i;
        }
    }
    return -1;
}

/**
 * @brief Validate that all characters in text are in alphabet
 */
static int validate_text(const char* text, const char* alphabet) {
    for (int i = 0; text[i]; i++) {
        if (find_char_index(alphabet, text[i]) == -1) {
            return -1;
        }
    }
    return 0;
}

/**
 * @brief Prepare text for processing
 */
static int prepare_text(const char* input, char* output, size_t output_size, const char* alphabet) {
    if (!input || !output || output_size == 0) {
        return -1;
    }
    
    size_t len = strlen(input);
    if (len >= output_size) {
        return -1;
    }
    
    strcpy(output, input);
    to_upper(output);
    
    // Remove characters not in alphabet
    int j = 0;
    for (int i = 0; output[i]; i++) {
        if (find_char_index(alphabet, output[i]) != -1) {
            output[j++] = output[i];
        }
    }
    output[j] = '\0';
    
    return 0;
}

/**
 * @brief Get shift value based on mode
 */
static int get_shift_value(
    const char* shift_mode,
    int shift_amount,
    int char_index,
    const int* custom_shifts,
    int custom_shifts_count,
    int* cumulative_shift
) {
    if (strcmp(shift_mode, "fixed") == 0) {
        return shift_amount;
    } else if (strcmp(shift_mode, "progressive") == 0) {
        *cumulative_shift += shift_amount;
        return *cumulative_shift;
    } else if (strcmp(shift_mode, "custom") == 0) {
        if (custom_shifts && char_index < custom_shifts_count) {
            return custom_shifts[char_index];
        }
        return 0;
    }
    return 0;
}

/**
 * @brief Encrypt text using the Reihenschieber cipher
 */
int reihenschieber_encrypt(
    const char* plaintext,
    const char* key,
    const char* alphabet,
    const char* shift_mode,
    const char* shift_direction,
    int shift_amount,
    const int* custom_shifts,
    int custom_shifts_count,
    char* result,
    size_t result_size
) {
    if (!plaintext || !key || !result || result_size == 0) {
        return -1;
    }
    
    if (!alphabet) {
        alphabet = DEFAULT_ALPHABET;
    }
    
    if (!shift_mode) {
        shift_mode = "fixed";
    }
    
    if (!shift_direction) {
        shift_direction = "forward";
    }
    
    char processed_text[MAX_TEXT_SIZE];
    char processed_key[MAX_TEXT_SIZE];
    
    if (prepare_text(plaintext, processed_text, sizeof(processed_text), alphabet) != 0) {
        return -1;
    }
    
    if (prepare_text(key, processed_key, sizeof(processed_key), alphabet) != 0) {
        return -1;
    }
    
    if (strlen(processed_text) >= result_size) {
        return -1;
    }
    
    int alphabet_len = strlen(alphabet);
    int key_len = strlen(processed_key);
    int text_len = strlen(processed_text);
    
    int key_index = 0;
    int cumulative_shift = 0;
    
    for (int i = 0; i < text_len; i++) {
        char current_char = processed_text[i];
        char current_key = processed_key[key_index % key_len];
        key_index++;
        
        int char_index = find_char_index(alphabet, current_char);
        int key_index_pos = find_char_index(alphabet, current_key);
        
        if (char_index == -1 || key_index_pos == -1) {
            return -1;
        }
        
        int current_shift = get_shift_value(shift_mode, shift_amount, i, custom_shifts, custom_shifts_count, &cumulative_shift);
        
        if (strcmp(shift_direction, "backward") == 0) {
            current_shift = -current_shift;
        }
        
        int encrypted_index = (char_index + key_index_pos + current_shift) % alphabet_len;
        if (encrypted_index < 0) {
            encrypted_index += alphabet_len;
        }
        
        result[i] = alphabet[encrypted_index];
    }
    result[text_len] = '\0';
    
    return 0;
}

/**
 * @brief Decrypt text using the Reihenschieber cipher
 */
int reihenschieber_decrypt(
    const char* ciphertext,
    const char* key,
    const char* alphabet,
    const char* shift_mode,
    const char* shift_direction,
    int shift_amount,
    const int* custom_shifts,
    int custom_shifts_count,
    char* result,
    size_t result_size
) {
    if (!ciphertext || !key || !result || result_size == 0) {
        return -1;
    }
    
    if (!alphabet) {
        alphabet = DEFAULT_ALPHABET;
    }
    
    if (!shift_mode) {
        shift_mode = "fixed";
    }
    
    if (!shift_direction) {
        shift_direction = "forward";
    }
    
    char processed_text[MAX_TEXT_SIZE];
    char processed_key[MAX_TEXT_SIZE];
    
    if (prepare_text(ciphertext, processed_text, sizeof(processed_text), alphabet) != 0) {
        return -1;
    }
    
    if (prepare_text(key, processed_key, sizeof(processed_key), alphabet) != 0) {
        return -1;
    }
    
    if (strlen(processed_text) >= result_size) {
        return -1;
    }
    
    int alphabet_len = strlen(alphabet);
    int key_len = strlen(processed_key);
    int text_len = strlen(processed_text);
    
    int key_index = 0;
    int cumulative_shift = 0;
    
    for (int i = 0; i < text_len; i++) {
        char current_char = processed_text[i];
        char current_key = processed_key[key_index % key_len];
        key_index++;
        
        int char_index = find_char_index(alphabet, current_char);
        int key_index_pos = find_char_index(alphabet, current_key);
        
        if (char_index == -1 || key_index_pos == -1) {
            return -1;
        }
        
        int current_shift = get_shift_value(shift_mode, shift_amount, i, custom_shifts, custom_shifts_count, &cumulative_shift);
        
        if (strcmp(shift_direction, "backward") == 0) {
            current_shift = -current_shift;
        }
        
        int decrypted_index = (char_index - key_index_pos - current_shift) % alphabet_len;
        if (decrypted_index < 0) {
            decrypted_index += alphabet_len;
        }
        
        result[i] = alphabet[decrypted_index];
    }
    result[text_len] = '\0';
    
    return 0;
}

/**
 * @brief Generate a random key for Reihenschieber cipher
 */
int reihenschieber_generate_random_key(int length, char* key, size_t key_size) {
    if (length <= 0 || !key || key_size == 0) {
        return -1;
    }
    
    if (length >= key_size) {
        return -1;
    }
    
    srand(time(NULL));
    
    for (int i = 0; i < length; i++) {
        key[i] = 'A' + (rand() % 26);
    }
    key[length] = '\0';
    
    return 0;
}

/**
 * @brief Generate a key of appropriate length for the given text
 */
int reihenschieber_generate_key_for_text(int text_length, char* key, size_t key_size) {
    if (text_length <= 0 || !key || key_size == 0) {
        return -1;
    }
    
    int key_length = 3 + (rand() % (text_length < 10 ? text_length - 2 : 8));
    return reihenschieber_generate_random_key(key_length, key, key_size);
}

/**
 * @brief Encrypt text with a randomly generated key
 */
int reihenschieber_encrypt_with_random_key(
    const char* plaintext,
    int key_length,
    char* result,
    size_t result_size,
    char* generated_key,
    size_t key_size
) {
    if (!plaintext || !result || !generated_key || result_size == 0 || key_size == 0) {
        return -1;
    }
    
    if (key_length <= 0) {
        key_length = 3 + (rand() % 8);
    }
    
    if (reihenschieber_generate_random_key(key_length, generated_key, key_size) != 0) {
        return -1;
    }
    
    return reihenschieber_encrypt(plaintext, generated_key, NULL, "fixed", "forward", 1, NULL, 0, result, result_size);
}

/**
 * @brief Produce custom shift patterns for Reihenschieber cipher
 */
int reihenschieber_produce_custom_shifts(
    const char* pattern_type,
    int pattern_length,
    int* shifts,
    size_t shifts_size
) {
    if (!pattern_type || pattern_length <= 0 || !shifts || shifts_size == 0) {
        return -1;
    }
    
    if (pattern_length >= shifts_size) {
        return -1;
    }
    
    if (strcmp(pattern_type, "alternating") == 0) {
        for (int i = 0; i < pattern_length; i++) {
            shifts[i] = (i % 2 == 0) ? 1 : -1;
        }
    } else if (strcmp(pattern_type, "fibonacci") == 0) {
        if (pattern_length >= 1) shifts[0] = 1;
        if (pattern_length >= 2) shifts[1] = 1;
        for (int i = 2; i < pattern_length; i++) {
            shifts[i] = shifts[i-1] + shifts[i-2];
        }
    } else if (strcmp(pattern_type, "prime") == 0) {
        int primes[] = {2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47};
        int primes_count = sizeof(primes) / sizeof(primes[0]);
        for (int i = 0; i < pattern_length; i++) {
            shifts[i] = primes[i % primes_count];
        }
    } else if (strcmp(pattern_type, "random") == 0) {
        srand(time(NULL));
        for (int i = 0; i < pattern_length; i++) {
            shifts[i] = (rand() % 11) - 5; // -5 to 5
        }
    } else {
        return -1;
    }
    
    return 0;
}

/**
 * @brief Encrypt Turkish text using Reihenschieber cipher
 */
int reihenschieber_encrypt_turkish(
    const char* plaintext,
    const char* key,
    const char* shift_mode,
    const char* shift_direction,
    int shift_amount,
    const int* custom_shifts,
    int custom_shifts_count,
    char* result,
    size_t result_size
) {
    return reihenschieber_encrypt(plaintext, key, TURKISH_ALPHABET, shift_mode, shift_direction, shift_amount, custom_shifts, custom_shifts_count, result, result_size);
}

/**
 * @brief Decrypt Turkish text using Reihenschieber cipher
 */
int reihenschieber_decrypt_turkish(
    const char* ciphertext,
    const char* key,
    const char* shift_mode,
    const char* shift_direction,
    int shift_amount,
    const int* custom_shifts,
    int custom_shifts_count,
    char* result,
    size_t result_size
) {
    return reihenschieber_decrypt(ciphertext, key, TURKISH_ALPHABET, shift_mode, shift_direction, shift_amount, custom_shifts, custom_shifts_count, result, result_size);
}

File: python/cryptology/classical/substitution/polyalphabetic/reihenschieber.py
"""
Reihenschieber Cipher Implementation

A mechanical polyalphabetic substitution cipher that uses shifting strips.
Essentially a mechanical Vigenère cipher with progressive shifting capabilities.

Features:
- Multiple shift modes: fixed, progressive, custom
- Shift directions: forward (default), backward
- Custom alphabets support (English, Turkish, etc.)
- Random key generation
- Self-reciprocal encryption/decryption
"""

import random
import string
from typing import List, Optional, Union


def reihenschieber_encrypt(
    plaintext: str,
    key: str,
    alphabet: Optional[str] = None,
    shift_mode: str = "fixed",
    shift_direction: str = "forward",
    shift_amount: int = 1,
    custom_shifts: Optional[List[int]] = None
) -> str:
    """
    Encrypt text using the Reihenschieber cipher.
    
    Args:
        plaintext: Text to encrypt
        key: Encryption key (repeats as needed)
        alphabet: Custom alphabet (default: English)
        shift_mode: "fixed", "progressive", or "custom"
        shift_direction: "forward" or "backward"
        shift_amount: Amount to shift (for fixed/progressive modes)
        custom_shifts: List of custom shift values (for custom mode)
    
    Returns:
        Encrypted text
    """
    if not plaintext or not key:
        return ""
    
    if alphabet is None:
        alphabet = string.ascii_uppercase
    
    # Prepare text and key
    text = plaintext.lower()
    key_upper = key.lower()
    alphabet_upper = alphabet.lower()
    
    # Validate inputs
    if not all(c in alphabet_upper for c in text):
        raise ValueError("Plaintext contains characters not in alphabet")
    if not all(c in alphabet_upper for c in key_upper):
        raise ValueError("Key contains characters not in alphabet")
    
    result = []
    key_index = 0
    cumulative_shift = 0
    
    for i, char in enumerate(text):
        if char not in alphabet_upper:
            result.append(char)
            continue
        
        # Get current key character
        current_key = key_upper[key_index % len(key_upper)]
        key_index += 1
        
        # Calculate shift based on mode
        if shift_mode == "fixed":
            current_shift = shift_amount
        elif shift_mode == "progressive":
            current_shift = shift_amount
            cumulative_shift += current_shift
        elif shift_mode == "custom":
            if custom_shifts is None or i >= len(custom_shifts):
                current_shift = 0
            else:
                current_shift = custom_shifts[i]
        else:
            raise ValueError(f"Invalid shift_mode: {shift_mode}")
        
        # Apply shift direction
        if shift_direction == "backward":
            current_shift = -current_shift
        
        # Calculate final shift
        if shift_mode == "progressive":
            final_shift = cumulative_shift
        else:
            final_shift = current_shift
        
        # Encrypt character
        char_index = alphabet_upper.index(char)
        key_index_pos = alphabet_upper.index(current_key)
        
        # Apply Vigenère-like encryption with additional shift
        encrypted_index = (char_index + key_index_pos + final_shift) % len(alphabet_upper)
        encrypted_char = alphabet_upper[encrypted_index]
        
        result.append(encrypted_char)
    
    return ''.join(result)


def reihenschieber_decrypt(
    ciphertext: str,
    key: str,
    alphabet: Optional[str] = None,
    shift_mode: str = "fixed",
    shift_direction: str = "forward",
    shift_amount: int = 1,
    custom_shifts: Optional[List[int]] = None
) -> str:
    """
    Decrypt text using the Reihenschieber cipher.
    
    Args:
        ciphertext: Text to decrypt
        key: Decryption key (repeats as needed)
        alphabet: Custom alphabet (default: English)
        shift_mode: "fixed", "progressive", or "custom"
        shift_direction: "forward" or "backward"
        shift_amount: Amount to shift (for fixed/progressive modes)
        custom_shifts: List of custom shift values (for custom mode)
    
    Returns:
        Decrypted text
    """
    if not ciphertext or not key:
        return ""
    
    if alphabet is None:
        alphabet = string.ascii_uppercase
    
    # Prepare text and key
    text = ciphertext.lower()
    key_upper = key.lower()
    alphabet_upper = alphabet.lower()
    
    # Validate inputs
    if not all(c in alphabet_upper for c in text):
        raise ValueError("Ciphertext contains characters not in alphabet")
    if not all(c in alphabet_upper for c in key_upper):
        raise ValueError("Key contains characters not in alphabet")
    
    result = []
    key_index = 0
    cumulative_shift = 0
    
    for i, char in enumerate(text):
        if char not in alphabet_upper:
            result.append(char)
            continue
        
        # Get current key character
        current_key = key_upper[key_index % len(key_upper)]
        key_index += 1
        
        # Calculate shift based on mode
        if shift_mode == "fixed":
            current_shift = shift_amount
        elif shift_mode == "progressive":
            current_shift = shift_amount
            cumulative_shift += current_shift
        elif shift_mode == "custom":
            if custom_shifts is None or i >= len(custom_shifts):
                current_shift = 0
            else:
                current_shift = custom_shifts[i]
        else:
            raise ValueError(f"Invalid shift_mode: {shift_mode}")
        
        # Apply shift direction
        if shift_direction == "backward":
            current_shift = -current_shift
        
        # Calculate final shift
        if shift_mode == "progressive":
            final_shift = cumulative_shift
        else:
            final_shift = current_shift
        
        # Decrypt character
        char_index = alphabet_upper.index(char)
        key_index_pos = alphabet_upper.index(current_key)
        
        # Apply Vigenère-like decryption with additional shift
        decrypted_index = (char_index - key_index_pos - final_shift) % len(alphabet_upper)
        decrypted_char = alphabet_upper[decrypted_index]
        
        result.append(decrypted_char)
    
    return ''.join(result)


def reihenschieber_generate_random_key(length: int) -> str:
    """
    Generate a random key for Reihenschieber cipher.
    
    Args:
        length: Length of the key to generate
    
    Returns:
        Random key string
    """
    if length <= 0:
        raise ValueError("Key length must be positive")
    
    return ''.join(random.choices(string.ascii_uppercase, k=length))


def reihenschieber_generate_key_for_text(text_length: int) -> str:
    """
    Generate a key of appropriate length for the given text.
    
    Args:
        text_length: Length of the text to encrypt
    
    Returns:
        Random key string
    """
    if text_length <= 0:
        raise ValueError("Text length must be positive")
    
    # Generate key length between 3 and min(text_length, 10)
    key_length = random.randint(3, min(text_length, 10))
    return reihenschieber_generate_random_key(key_length)


def reihenschieber_encrypt_with_random_key(
    plaintext: str,
    key_length: Optional[int] = None
) -> tuple[str, str]:
    """
    Encrypt text with a randomly generated key.
    
    Args:
        plaintext: Text to encrypt
        key_length: Length of key to generate (default: auto)
    
    Returns:
        Tuple of (encrypted_text, generated_key)
    """
    if not plaintext:
        return "", ""
    
    if key_length is None:
        key_length = random.randint(3, min(len(plaintext), 10))
    
    generated_key = reihenschieber_generate_random_key(key_length)
    encrypted = reihenschieber_encrypt(plaintext, generated_key)
    
    return encrypted, generated_key


def reihenschieber_produce_custom_shifts(
    pattern_type: str = "alternating",
    pattern_length: int = 10,
    **kwargs
) -> List[int]:
    """
    Produce custom shift patterns for Reihenschieber cipher.
    
    Args:
        pattern_type: Type of pattern ("alternating", "fibonacci", "prime", "random")
        pattern_length: Length of the pattern
        **kwargs: Additional parameters for specific patterns
    
    Returns:
        List of shift values
    """
    if pattern_length <= 0:
        raise ValueError("Pattern length must be positive")
    
    if pattern_type == "alternating":
        # Alternating positive and negative shifts
        return [1 if i % 2 == 0 else -1 for i in range(pattern_length)]
    
    elif pattern_type == "fibonacci":
        # Fibonacci-like pattern
        shifts = [1, 1]
        for i in range(2, pattern_length):
            shifts.append(shifts[i-1] + shifts[i-2])
        return shifts[:pattern_length]
    
    elif pattern_type == "prime":
        # Prime number pattern
        primes = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47]
        return [primes[i % len(primes)] for i in range(pattern_length)]
    
    elif pattern_type == "random":
        # Random shifts between -5 and 5
        return [random.randint(-5, 5) for _ in range(pattern_length)]
    
    else:
        raise ValueError(f"Invalid pattern_type: {pattern_type}")


# Turkish alphabet support
TURKISH_ALPHABET = "ABCÇDEFGĞHIİJKLMNOÖPRSŞTUÜVYZ"


def reihenschieber_encrypt_turkish(
    plaintext: str,
    key: str,
    shift_mode: str = "fixed",
    shift_direction: str = "forward",
    shift_amount: int = 1,
    custom_shifts: Optional[List[int]] = None
) -> str:
    """
    Encrypt Turkish text using Reihenschieber cipher.
    
    Args:
        plaintext: Turkish text to encrypt
        key: Encryption key
        shift_mode: "fixed", "progressive", or "custom"
        shift_direction: "forward" or "backward"
        shift_amount: Amount to shift
        custom_shifts: List of custom shift values
    
    Returns:
        Encrypted Turkish text
    """
    return reihenschieber_encrypt(
        plaintext, key, TURKISH_ALPHABET, shift_mode, 
        shift_direction, shift_amount, custom_shifts
    )


def reihenschieber_decrypt_turkish(
    ciphertext: str,
    key: str,
    shift_mode: str = "fixed",
    shift_direction: str = "forward",
    shift_amount: int = 1,
    custom_shifts: Optional[List[int]] = None
) -> str:
    """
    Decrypt Turkish text using Reihenschieber cipher.
    
    Args:
        ciphertext: Turkish text to decrypt
        key: Decryption key
        shift_mode: "fixed", "progressive", or "custom"
        shift_direction: "forward" or "backward"
        shift_amount: Amount to shift
        custom_shifts: List of custom shift values
    
    Returns:
        Decrypted Turkish text
    """
    return reihenschieber_decrypt(
        ciphertext, key, TURKISH_ALPHABET, shift_mode, 
        shift_direction, shift_amount, custom_shifts
    )
